### PR TITLE
Upgrades to PSym

### PR DESCRIPTION
**Pull request by aman-goel**
Hi,

I have added several upgrades to PSym. Here is a summary of major changes:

- Adds state-space coverage reporting
- Adds foreach to while transformer
- Add support for specifying test method
- Support map value summary with tuple keys
- Adds initial support for allowing recursive functions
- Adds MonoSAT backend
- Adds memory/time monitors
- Updates loggers, scripts, and readme
- Updates storing state in file and resuming

Ankush (@ankushdesai), it would be great if you can review and let me know of any suggested changes.

--
Aman
@aman-goel

### CODE REVIEW

File: Src/PRuntimes/PSymbolicRuntime/pom.xml [hunk 1]
```diff
@@ -260,22 +285,22 @@
...
+      <artifactId>log4j</artifactId>
+      <version>1.2.17</version>
     </dependency>
 
-    <dependency>
-      <groupId>org.apache.logging.log4j</groupId>
-      <artifactId>log4j-core</artifactId>
-      <version>[2.17.1,)</version>
-    </dependency>
+<!--    <dependency>-->
```
[the remark this answers is not in the document]
Comment (ankushdesai) [reply Src/PRuntimes/PSymbolicRuntime/pom.xml:293]: I would recommend putting the different solvers and installing the dependencies as optional. It should build and work even without getting the solvers.

**Review by ankushdesai — CHANGES_REQUESTED** [review]

File: Src/PCompiler/CompilerCore/Backend/Symbolic/SymbolicCodeGenerator.cs [hunk 2]
```diff
@@ -926,25 +968,9 @@ private void WriteStmt(Function function, CompilationContext context, StringWrit
...
-                            }
-                            context.Write(output, ");");
-                            break;
-                        default:
-                            context.Write(output, "(");
-                            WriteExpr(context, output, flowContext.pcScope, printStmt.Message);
-                            context.Write(output, ").toString());");
-                            break;
-                    }
+                    context.Write(output, "System.out.println(");
```
[the remark this answers is not in the document]
Comment (aman-goel) [reply Src/PCompiler/CompilerCore/Backend/Symbolic/SymbolicCodeGenerator.cs]: Resolved in 85b7094df133cc6eef915856b394110d79419f79

File: Bld/build.sh [hunk 3]
```diff
@@ -16,6 +16,10 @@ git submodule update --init --recursive
 echo -e "${ORANGE} ---- Building the Java P runtime ----${NOCOLOR}"
 mvn compile -q -f ./Src/PRuntimes/PJavaRuntime/pom.xml
 
+echo -e "${ORANGE} ---- Building the PSym runtime ----${NOCOLOR}"
```
[the remark this answers is not in the document]
Comment (aman-goel) [reply Bld/build.sh]: Resolved in 05bc2c159281ee019336a78f741a120ecb74b9ee

File: Src/PRuntimes/PSymbolicRuntime/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ankushdesai) [reply Src/PRuntimes/PSymbolicRuntime/pom.xml:293]: @aman-goel, one recommendation I have is to have other solver options as hidden options that are not visible to the users.

File: Src/PCompiler/CompilerCore/Backend/Symbolic/TransformASTPass.cs [hunk 4]
```diff
@@ -551,61 +569,62 @@ static private IPStmt HandleReceives(IPStmt statement, Function function, Machin
...
-                                 if (after != null)
-                                 {
-                                     foreach (var stmt in after.Statements)
-                                     {
-                                         result.Add(ReplaceVars(stmt, newVarMap));
-                                     }
-                                 }
-                                 whileNumber++;
-                                 break;
+                                 throw new NotImplementedException($"Receive in a while statement is not yet supported, found in {machine.Name}");
```
Comment (ankushdesai) [inline Src/PCompiler/CompilerCore/Backend/Symbolic/TransformASTPass.cs:572]: Can I request you to create a issue on github that adds a todo that there is a potential bug in the receive in while logic alone with a small repro for the bug that @lmpick can look at. Thanks!

File: Src/PCompiler/CompilerCore/Compiler.cs [hunk 5]
```diff
@@ -31,6 +31,7 @@ public void Compile(ICompilationJob job)
             Scope scope = Analyzer.AnalyzeCompilationUnit(job.Handler, trees);
 
             // Convert functions to lowered SSA form with explicit cloning
+            IRTransformer.SetOutputLanguage(job.OutputLanguage);
```
Comment (ankushdesai) [inline Src/PCompiler/CompilerCore/Compiler.cs]: I dont think we need this anymore. Can we drop this line as we dont do language specific transformations.

Comment (ankushdesai) [reply Src/PCompiler/CompilerCore/Compiler.cs]: Also, drop the `SetOutputLanguage` function.

**Review by ankushdesai — CHANGES_REQUESTED** [review]

File: Src/PCompiler/CompilerCore/Backend/Symbolic/TransformASTPass.cs (hunk 4 above)
Comment (aman-goel) [reply Src/PCompiler/CompilerCore/Backend/Symbolic/TransformASTPass.cs:572]: Added: Check issue #501

File: Src/PCompiler/CompilerCore/Compiler.cs (hunk 5 above)
Comment (aman-goel) [reply Src/PCompiler/CompilerCore/Compiler.cs]: (Already) resolved in 1b33a4628d38b61cc2deb68ec6314925fa72855b

**Review by ankushdesai — APPROVED** [review]
Thanks for the changes!

**Pull request merged**